Draft: adapt authz docs to introduction of user management

meiersi-da opened the pull request:
Fixes #12348 and #12347.

Open tasks:
- [x] get confirmation from @bame-da and @gerolf-da that it is OK to remove the services diagram on https://docs.daml.com/app-dev/services.html#services-diagram, as there is not a single link to it
- [x] polish docs for user management api services
- [x] fix access rights docs for user management on http://127.0.0.1:8000/app-dev/authorization.html#authorization-claims
- [x] remove redundant section on token formats for sandbox
- [x] replace duplicate table of claims in `Claims.scala` with forward pointer to the docs
- [x] review all other tools sections wrt their relation to the new authorization docs

Docs left for a later update by component owners:
- #12439
- #12440
- #12441

CHANGELOG_BEGIN
- [User management]: document user management service on https://docs.daml.com/app-dev/services.html and user access tokens on https://docs.daml.com/app-dev/authorization.html
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description
- [x] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

gerolf-da commented:
I'm fine with removing it.

bame-da commented:
Also fine with me, but I defer product opinion on docs matters like this to @carrielaben-da

meiersi-da commented:
> Also fine with me, but I defer product opinion on docs matters like this to @carrielaben-da
> 
> Ed: Since it's a vacation in the US, if you want to get this merged, I'd go ahead and discuss with @carrielaben-da later if needed.

Thanks for the feedback @bame-da . The PR is now ready for review. Would you have time for reviewing (or delegating the review) wrt whether the content matches the design we discussed? 

@carrielaben-da would you have time to do a review as well?

Many thanks :pray:

meiersi-da commented:
@mziolekda @cocreature @stefanobaghino-da could I ask you for a (cursory) review as well, so we are sure that we're all in sync on the user management and user access tokens?

cocreature commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
+   instead of being nested below ``"https://daml.com/ledger-api"``.
+
+
+User access tokens
```
should we move user tokens above the custom tokens to make that the thing people reach for first?

meiersi-da replied:
good call. done.

cocreature commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
 #############
```
Maybe this is not the right section for this but reading through it it seemed to be missing some instructions for how you actually use this on the gRPC API or the JSON API. Doesn’t have to be directly in here but I think at least some pointers would be very helpful.

meiersi-da replied:
Good point. I've added this paragraph here: https://github.com/digital-asset/daml/pull/12398/commits/be09ee4d8a7fb57f34ea52b8fb98bda81dff604d

I'd hope that @stefanobaghino-da and his team can add further pointers in later PRs as part of their work on the docs for the JSON-API, auth-middleware, and other tools that are being adapted.

cocreature reviewed (approved):
Nice work, thank you!

meiersi-da commented on docs/scripts/check-closing-quotes.sh.allow:
````diff
@@ -4,3 +4,4 @@
 _\\ \\/ _ `/ _ \\/ _  / _ \\/ _ \\\\ \\ /
 /    / _ `/ |/ / / _ `/ _ `/ __/ _ \\/ __/
 -- ```
+\^\$\.!`-#+
````
This was to fix the error from the `check-closing-quotes.sh` script complaining about: https://github.com/digital-asset/daml/pull/12398/files#diff-c58c7f4b4aee829d88d36290b03495a358d28898cfe9ecb9b2556a552f8fd580R55

meiersi-da replied:
The problem is the unclosed '`'

meiersi-da commented on docs/source/app-dev/services.rst:
```diff
@@ -272,8 +290,3 @@ Use the **reset service** to reset the ledger state, as a quicker alternative to
-Services diagram
-****************
-
-.. image:: ./images/services.svg
```
good catch. done.

carrielaben-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
 Introduction
 ************
 
-The :doc:`Ledger API </app-dev/ledger-api>` is used to request changes to the ledger (e.g., "*Alice
```
"ledger API" (capitalization)

meiersi-da replied:
https://docs.daml.com/concepts/glossary.html#ledger-api is the current definition that I've used. The capitalization of "Ledger API" makes sense to as it is a standalone concept.

If you'd like to change it, then I'd suggest to do it as a search-and-replace in a separate PR to ensure consistency.

carrielaben-da replied:
No worries, if there's a conceptual reason for it to be capitalized here that works.

carrielaben-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
-The :doc:`Ledger API </app-dev/ledger-api>` is used to request changes to the ledger (e.g., "*Alice
-wants to exercise choice X on contract Y*"), or to read data from the ledger (e.g., "*Alice wants to
-see all active contracts*").
+Your Daml application sends requests to the :doc:`Ledger API </app-dev/ledger-api>` exposed by a participant node to submit changes to the ledger
```
"ledger API" (capitalization)

carrielaben-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
-and does not require any cryptographic information to be sent along with requests.
+Whether a participant node *can* serve such a request depends on whether the participant node hosts the respective parties, and
+whether the request is valid according to the :ref:`Daml Ledger Model <da-ledgers>`.
+Whether a participant node *will* serve such a request to a Daml application depends on whether the application accompanied the
```
A little loquacious - might be better as "whether the request includes an access token" or something similar.

meiersi-da replied:
thx. addressed.

carrielaben-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
-How authentication is set up on a particular ledger is defined by the ledger operator.
-However, most authentication setups share the following pattern:
+How an application should acquire access tokens depends on the participant node it talks to and is ultimately setup by the participant node operator.
+However most setups share the following pattern:
```
Strike "However"

meiersi-da replied:
done

carrielaben-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
-(e.g., by checking the username/password credentials sent with the request),
-looks up the privileges of the user,
+The token issuer verifies the identity of the requesting application
+(e.g., by checking the `OAuth 2.0 client credentials <https://datatracker.ietf.org/doc/html/rfc6749#section-2.2>`_ sent with the request),
```
Does the person reading the documentation need to know how it's done, or only that it's done?

meiersi-da replied:
Addressed.

meiersi-da commented:
@carrielaben-da : I've addressed your review comments except for the one on the Ledger API. Could have a look at my answer there and approve the PR or request further changes?

carrielaben-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
+- ``canReadAs(p)``: the right to read information off the ledger (like the active contracts) visible to the party ``p``
+- ``canActsAs(p)``: same as ``canReadAs(p)``, with the added right of issuing commands on behalf of the party ``p``
+
+The following table summarizes what kind of rights are required to access each Ledger API endpoint:
```
"what kind of rights are required" -> "the rights required"

carrielaben-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,189 @@ Authorization
+They are useful for scenarios where an application's rights change dynamically over the application's lifetime.
+
+User access tokens do not encode rights directly like custom Daml claims tokens.
+Instead, user access tokens encode the participant user on behalf of which the application is issuing a request.
```
"Instead, these tokens encode the participant user on whose behalf the request is issued."

carrielaben-da commented on docs/source/app-dev/services.rst:
```diff
@@ -193,7 +193,25 @@ Party management service
 Use the **party management service** to allocate parties on the ledger and retrieve information about allocated parties.
 
-Allocating parties is necessary to interact with the ledger. For more information, refer to the pages on :doc:`Identity Management</concepts/identity-and-package-management>` and :ref:`the API reference documentation <com.daml.ledger.api.v1.admin.PartyManagementService>`.
+Parties serve to govern on-ledger access control as per :ref:`Daml's privacy model <da-model-privacy>`
```
"Parties govern on-ledger access..."

carrielaben-da commented on docs/source/app-dev/services.rst:
```diff
@@ -193,7 +193,25 @@ Party management service
-Allocating parties is necessary to interact with the ledger. For more information, refer to the pages on :doc:`Identity Management</concepts/identity-and-package-management>` and :ref:`the API reference documentation <com.daml.ledger.api.v1.admin.PartyManagementService>`.
+Parties serve to govern on-ledger access control as per :ref:`Daml's privacy model <da-model-privacy>`
+and :ref:`authorization rules <da-ledgers-authorization-rules>`.
+Applications and their operators are expected allocate and use parties to manage on-ledger access control as per their business requirements.
```
"are expected to allocate and use..."

carrielaben-da commented on docs/source/json-api/index.rst:
```diff
@@ -267,7 +267,7 @@ The JSON API essentially performs two separate tasks:
 ----------------------------
 
-Party-specific requests, i.e., command submissions and queries, require a JWT with some additional restrictions compared to the format :doc:`described in the Token Payload section here </tools/sandbox>`. For command submissions, ``actAs`` must contain at least one party and ``readAs`` can contain 0 or more parties. Queries require at least one party in either ``actAs`` or ``readAs`` (note that be ...
+Party-specific requests, i.e., command submissions and queries, require a JWT with some additional restrictions compared to general :ref:`Daml ledger access token formats <access-token-formats>`. For command submissions, ``actAs`` must contain at least one party and ``readAs`` can contain 0 or more parties. Queries require at least one party in either ``actAs`` or ``readAs`` (note that before SDK ...
```
"In addition to that" --> "In addtion,"

carrielaben-da replied:
"The application id and ledger id are also mandatory" is even shorter.

bame-da replied:
I would elaborate this paragraph a little more. @stefanobaghino-da maybe you can make a suggestion here or open a ticket to do so later. In short, I think it should be:

When running against a ledger that supports user management, the JSON API accepts any user token and uses that to resolve the bearers rights.

When using custom claims tokens, there is a restriction that at least one party must be set - in `actAs` for submissions, or anywhere for read requests.

meiersi-da replied:
@bame-da (FYI: @stefanobaghino-da) added your comment to https://github.com/digital-asset/daml/issues/12440

meiersi-da replied:
@carrielaben-da : addressed your comments in an upcoming commit

carrielaben-da commented on ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/user_management_service.proto:
```diff
@@ -56,48 +46,46 @@ service UserManagementService {
-    // Ledger API clients MAY set this field to empty for special users; e.g., a user
-    // that is granted ``CanReadAsAnyParty`` so that it can export the data for all parties
-    // hosted on the participant node.
+    // Users for participant administrators MAY not have an associated primary party.
```
A little unclear - does this capitalization indicate 'must not' or 'may or may not'?

bame-da replied:
may or may not

meiersi-da replied:
fixed

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
-The token issuer verifies the identity of the requesting user
-(e.g., by checking the username/password credentials sent with the request),
-looks up the privileges of the user,
+The token issuer verifies the identity of the requesting application looks up the privileges of the application,
```
```suggestion
The token issuer verifies the identity of the requesting application, looks up the privileges of the application,
```

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
-The Daml ledger verifies the signature of the token (to make sure it has not been tampered with),
-and then checks that the privileges described in the token authorize the given ledger API request.
+Then, the Daml application sends the access token along with every ledger API request.
+The Daml ledger verifies the signature of the token (to make sure it has not been tampered with and was issued by one of its trusted token issuers),
```
```suggestion
The Daml ledger verifies the signature of the token to make sure it has not been tampered with and was issued by one of its trusted token issuers,
```

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
-and then checks that the privileges described in the token authorize the given ledger API request.
+Then, the Daml application sends the access token along with every ledger API request.
+The Daml ledger verifies the signature of the token (to make sure it has not been tampered with and was issued by one of its trusted token issuers),
+and then checks that the token has not yet expired and that the privileges described in the token authorize the given ledger API request.
```
```suggestion
and then checks that the token has not yet expired and that the privileges described in the token authorize the given Ledger API request.
```

;)

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
 
-Glossary:
+As shown above, using access tokens requires your application to attach them to every request.
+How to do that depends on the tool or library you use to interact with the ledger API.
```
That is true, but I think we should describe the mechanism at the gRPC layer.
Ie something like "the Ledger API expects a trailer of the format here demonstrated with a gRPCurl command: "

meiersi-da replied:
Will do in https://github.com/digital-asset/daml/issues/12501

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
+User access tokens
+==================
+
+Daml ledger that support participant user management also accept user access tokens.
```
```suggestion
Daml ledgers that support participant user management also accept user access tokens.
```

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
+Daml ledger that support participant user management also accept user access tokens.
+They are useful for scenarios where an application's rights change dynamically over the application's lifetime.
+
+User access tokens do not encode rights directly like custom Daml claims tokens.
```
Forward reference to the custom Daml claims tokens header here.

meiersi-da replied:
Overlooked that one. Fixing in another PR.

meiersi-da replied:
https://github.com/digital-asset/daml/pull/12534

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
+- ``exp`` is the standard JWT expiration date (in seconds since EPOCH)
+- ``actAs``, ``readAs`` and (participant) ``admin`` encode the rights granted by this access token
+
+The ``public`` right is implicitly granted to any bearing a valid JWT issued by a trusted issuer (even without being an admin or being able to act or read on behalf of any party).
```
```suggestion
The ``public`` right is implicitly granted to anyone bearing a valid JWT issued by a trusted issuer (even without being an admin or being able to act or read on behalf of any party).
```

Is this true? Don't we even check participantId or applicationId or anything? Could I use an empty token or a token issued for a completely different audience by the same token issuer?

meiersi-da replied:
> Is this true? Don't we even check participantId or applicationId or anything? Could I use an empty token or a token issued for a completely different audience by the same token issuer?

We do check not expired, ledgerId, and participantId for a token to be considered valid: https://github.com/digital-asset/daml/blob/2deeabf3fac6ec6d75f9d779f4acde1c0dacc56b/ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala#L43

Moreover, every request that does specify an `applicationId` compares that one to the one in the token if set.

I'll clarify the text in a separate PR.

meiersi-da replied:
https://github.com/digital-asset/daml/pull/12534

bame-da commented on docs/source/app-dev/authorization.rst:
```diff
@@ -7,124 +7,195 @@ Authorization
+The ``public`` right is implicitly granted to any bearing a valid JWT issued by a trusted issuer (even without being an admin or being able to act or read on behalf of any party).
+
+.. note:: All Daml ledgers also support a deprecated legacy format of custom Daml claims
+   access tokens whose format is equal to the above expect for the custom claims
```
```suggestion
   access tokens whose format is equal to the above except for the custom claims
```

meiersi-da replied:
https://github.com/digital-asset/daml/pull/12534/files

bame-da commented on docs/source/app-dev/services.rst:
```diff
@@ -193,7 +193,25 @@ Party management service
+User management service
+=======================
+
+Use the **user management service** to manage the set of users on a participant node and their :ref:`access rights <authorization-claims>` to that node's ledger API services.
```
```suggestion
Use the **user management service** to manage the set of users on a participant node and their :ref:`access rights <authorization-claims>` to that node's Ledger API services.
```
:P

bame-da replied:
Going to stop flagging it up here. Just need to make a decision whether to capitalize or not and go with one or the other.

meiersi-da replied:
I'll search an replace to capitalized to keep the current style in the glossary.

bame-da commented on docs/source/tools/sandbox.rst:
```diff
@@ -132,39 +133,11 @@ use one of the following command line options:
-
-The ``public`` claim is implicitly held by anyone bearing a valid JWT (even without being an admin or being able to act or read on behalf of any party).
 
 Generating JSON Web Tokens (JWT)
```
Wondering whether this would better sit on the auth page rather than the Sandbox page.

meiersi-da replied:
I've added a corresponding note to the authorization.rst page, but left this section on the sandbox page to keep the flow.

bame-da commented on ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/user_management_service.proto:
```diff
@@ -56,48 +46,46 @@ service UserManagementService {
-    // Ledger API clients MAY set this field to empty for special users; e.g., a user
-    // that is granted ``CanReadAsAnyParty`` so that it can export the data for all parties
-    // hosted on the participant node.
+    // Users for participant administrators MAY not have an associated primary party.
```
```suggestion
    // Users for participant administrators MAY have an associated primary party.
```

meiersi-da replied:
Ah good point. I'll change it to this wording.

bame-da reviewed (approved):
Left a few small comments. Didn't spend time copy-editing, just picked up what my eyes stumbled upon.

In terms of content for reference documentation, I think this is sufficiently complete with respect to the Ledger API.
